src/models/whisper.cpp

```diff
@@ -66,21 +252,174 @@ RoamingArray<float> Whisper_State::Run(int current_length, RoamingArray<int32_t>
+
+    auto output_cross_qk_dims = output_cross_qk_[0]->GetTensorTypeAndShapeInfo()->GetShape();
+    cuda::LaunchCopyCrossQKSingleDecodeStep(model_.cuda_stream_,
+                                            cross_qk_search_buffer_->GetTensorMutableData<float>(),
+                                            output_cross_qk_ptrs_gpu_.data(),
+                                            current_length - params_->sequence_length,
+                                            params_->BatchBeamSize(),
+                                            model_.config_->model.decoder.num_hidden_layers,
+                                            static_cast<int32_t>(output_cross_qk_dims[1]),
+                                            static_cast<int32_t>(alignment_heads_->GetTensorTypeAndShapeInfo()->GetShape()[0]),
+                                            alignment_heads_->GetTensorData<int32_t>(),
+                                            static_cast<int32_t>(output_cross_qk_dims[3]),
+                                            params_->search.max_length);
+#endif
+  }
+}
+
+void Whisper_State::Finalize() {
+  if (output_cross_qk_.size() && alignment_heads_) {
+#if USE_CUDA
+    int decoded_length = *(past_sequence_length_->GetTensorMutableData<int32_t>()) + 1;
+    auto output_cross_qk_dims = output_cross_qk_[0]->GetTensorTypeAndShapeInfo()->GetShape();
+
+    // Instantiate final output for cross QKs
+    auto num_alignment_heads = alignment_heads_->GetTensorTypeAndShapeInfo()->GetShape()[0];
+    auto cross_qk_type = model_.session_info_->GetOutputDataType("output_cross_qk_0");
+    auto cross_qk_shape = std::array<int64_t, 5>{params_->batch_size, params_->search.num_return_sequences, num_alignment_heads, decoded_length, 1500};
+    cross_qk_final_ = OrtValue::CreateTensor(*model_.allocator_device_, cross_qk_shape, cross_qk_type);
+
+    cuda::LaunchFinalizeCrossQK(model_.cuda_stream_,
+                                decoded_length - params_->sequence_length,
+                                decoded_length,
+                                static_cast<int32_t>(output_cross_qk_dims[0]),
+                                params_->search.num_beams,
+                                params_->search.max_length,
+                                static_cast<int32_t>(alignment_heads_->GetTensorTypeAndShapeInfo()->GetShape()[0]),
+                                static_cast<int32_t>(output_cross_qk_dims[3]),
+                                cross_qk_search_buffer_->GetTensorData<float>(),
+                                cross_qk_final_->GetTensorMutableData<float>(),
+                                params_->search.num_return_sequences,
+                                cache_indirection_->GetTensorData<int32_t>());
+#endif
+  }
 }
 
+OrtValue* Whisper_State::GetOutput(const char* name) {
+  if (std::strcmp("cross_qk", name) == 0) {
```

Comment:
>"cross_qk" [](http://example.com/codeflow?start=18&length=10)

could we remove this dirty hard code?

Reply:
Added some code comments highlighting that the cross_qk_final_ is a buffer maintained by ort-genai.

We need to document it in an example and probably in our API docs if possible. We can do this in a follow up pull-request.